### PR TITLE
Add return type for Eloquent Factories

### LINKED ISSUE #661: Support Laravel 8 factories
- Larastan Version: 0.6.4
- `--level` used: max

### Description
It's seems that states on the new factories that came with Laravel 8 are not yet supported.

It would be nice if we could deduct the return type of the `Model::factory()` method, added by the [HasFactory](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Factories/HasFactory.php) trait.

### Laravel code where the issue was found

```php
User::factory()->expired()->create();
```

I would love to help out, but I haven't got a clue where to start. If anyone could guide me, I'm very much willing to put some time and effort into this.

### PR DESCRIPTION
**Pull request by robbplo**
- [ ] Added or updated tests
- [ ] Documented user facing changes
- [ ] Updated CHANGELOG.md

Resolves #661

_Will update with more info soon_

Test is still failing, proper return type is included but in union type with basic factory class, not sure how to resolve this yet.
```
There was 1 failure:

1) Tests\FeaturesTest::testFeatures with data set "ReturnTypes/ModelFactoryReturnTypeTest.php" ('/Users/robbinploeger/larastan...st.php')
{
    "totals": {
        "errors": 0,
        "file_errors": 1
    },
    "files": {
        "\/Users\/robbinploeger\/larastan\/tests\/Features\/ReturnTypes\/ModelFactoryReturnTypeTest.php": {
            "errors": 1,
            "messages": [
                {
                    "message": "Method Tests\\Features\\ReturnTypes\\ModelFactoryReturnTypeTest::testFactoryCallFromModelReturnsSpecificFactory() should return Tests\\Application\\database\\factories\\UserFactory but returns Database\\Factories\\UserFactory|Illuminate\\Database\\Eloquent\\Factories\\Factory.",
                    "line": 13,
                    "ignorable": true
                }
            ]
        }
    },
    "errors": []
}
```

**Changes**

**Breaking changes**

### CODE REVIEW

**Comment by canvural** [comment]
Hi,

Thank you for your time for this. I didn't review it completely because you marked it as a draft. But there is one thing I noticed. We need to support Laravel 6.x, 7.x, and 8.x  Adding the `HasFactory` trait to `User` model breaks the tests in other Laravel versions because the trait does not exist. We need to find a solution for that.

File: tests/Features/ReturnTypes/ModelFactoryReturnTypeTest.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Tests\Features\ReturnTypes;
+
+use App\User;
+use Tests\Application\database\factories\UserFactory;
+
+class ModelFactoryReturnTypeTest
+{
+    /** @phpstan-return UserFactory */
+    public function testFactoryCallFromModelReturnsSpecificFactory()
```
Comment (szepeviktor) [inline tests/Features/ReturnTypes/ModelFactoryReturnTypeTest.php:11]: How about this?

```suggestion
    public function testFactoryCallFromModelReturnsSpecificFactory(): UserFactory
```

**Comment by Daanra** [comment]
> Test is still failing, proper return type is included but in union type with basic factory class, not sure how to resolve this yet.

@robbplo The union is caused by another Extension which also matches the Model::factory() method. You can edit `NunoMaduro\Larastan\ReturnTypes\ModelExtension::isStaticMethodSupported` to prevent that extension from interfering with yours.

**Comment by Daanra** [comment]
@robbplo Any update on this? If you're not planning on finishing this, I'll open another PR myself.

**Comment by canvural** [comment]
@Daanra Did you find a way to make the tests also pass for Laravel 6.x?

**Comment by robbplo** [comment]
@Daanra I haven't been able to figure this one out. Hope you can solve it!

**Comment by Daanra** [comment]
@canvural I was thinking of altering the testing script a bit such that certain files will only be tested on a specific Laravel version.

**Pull request closed without merging**